### PR TITLE
OSDOCS-6894:Added ControlPlaneMachineSet support for vSphere

### PR DESCRIPTION
**Pull request by dfitzmau**
[OSDOCS-6894](https://issues.redhat.com/browse/OSDOCS-6894)

Version(s):
4.15 (release notes TP table update required)

Link to docs preview:
* [Supported cloud providers](https://68190--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-getting-started#cpmso-platform-matrix_cpmso-getting-started)
* [Sample VMware vSphere failure domain configuration](https://68190--ocpdocs-pr.netlify.app/openshift-enterprise/latest/machine_management/control_plane_machine_management/cpmso-configuration#cpmso-yaml-failure-domain-vsphere_cpmso-configuration)

SME review:
- [x] SME has approved this change.

QE review:
- [x] QE has approved this change.

Additional information:
* [Git Gist](https://gist.github.com/rvanderp3/c74a19970d4878cc1da820e3d763495b)
* [GH update](https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/267/files)

### CODE REVIEW

File: machine_management/control_plane_machine_management/cpmso-getting-started.adoc [hunk 1]
```diff
@@ -63,6 +63,7 @@ The status of the control plane machine set after installation depends on your c
 1. AWS clusters that are upgraded from version 4.11 or earlier require xref:../../machine_management/control_plane_machine_management/cpmso-getting-started.adoc#cpmso-activating_cpmso-getting-started[CR activation].
 2. GCP and Azure clusters that are upgraded from version 4.12 or earlier require xref:../../machine_management/control_plane_machine_management/cpmso-getting-started.adoc#cpmso-activating_cpmso-getting-started[CR activation].
 3. Nutanix and {rh-openstack} clusters that are upgraded from version 4.13 or earlier require xref:../../machine_management/control_plane_machine_management/cpmso-getting-started.adoc#cpmso-activating_cpmso-getting-started[CR activation].
+4. Installing a cluster on VWware vSphere, includes a generated `ControlPlaneMachineSet` CR and activates the CR by default. This is a Technology Preview feature for VMware vSphere. See link:https://access.redhat.com/support/offerings/techpreview[Technology Preview Features - Scope of Support].
```
Comment (dfitzmau) [inline machine_management/control_plane_machine_management/cpmso-getting-started.adoc:66]: This TP feature that is applied automatically after installing a cluster means that our customers are automatically opted in to a TP feature. but would they typically have to manually opt in to a TP feature given the level of support?

Comment (dfitzmau) [reply machine_management/control_plane_machine_management/cpmso-getting-started.adoc:66]: The UPI vSphere docs include a link to [Creating a compute machine set on vSphere](https://docs.openshift.com/container-platform/4.14/machine_management/creating_machinesets/creating-machineset-vsphere.html#creating-machineset-vsphere_creating-machineset-vsphere](https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installing-vsphere.html#installation-machine-requirements_installing-vsphere)

The Supported cloud providers section is embedded in the Machine management docs, but should I include a link in the vSphere UPI and IPI docs? Then agin, is the TP feature only for IPI? For UPI, a customer would need to generate the `ControlPlaneMachineSet` CR and then activate?

File: modules/cpmso-yaml-failure-domain-vsphere.adoc [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// Module included in the following assemblies:
+//
+// * machine_management/cpmso-configuration.adoc
+
+:_mod-docs-content-type: REFERENCE
+[id="cpmso-yaml-failure-domain-vsphere_{context}"]
+= Sample VMware vSphere failure domain configuration
+
+On VMware vSphere infrastructure, a failure domain gets defined in an infrastructure resource specification. A failure domain is an infrastructure resource that comprises a control plane machine set, a vCenter datacenter, vCenter datastore, and a network. By using a failure domain resource, you can deploy control plane machines on hardware that is separate from the primary VMware vSphere infrastructure.
```
Comment (dfitzmau) [inline modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: Query to SME: The infrastructure resource specification is equivalent to the "provider specification"? See   [Sample install-config.yaml file for an installer-provisioned VMware vSphere cluster](https://docs.openshift.com/container-platform/4.14/machine_management/control_plane_machine_management/cpmso-configuration.html#cpmso-yaml-provider-spec-vsphere_cpmso-configuration)

Comment (rvanderp3) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: not quite, the infrastructure resource defines the failure domains, but the cpms spec specifies the names of the failure domains to use.

File: modules/cpmso-yaml-failure-domain-vsphere.adoc [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+:FeatureName: Defining a failure domain for a control plane machine set
+include::snippets/technology-preview.adoc[]
+
+// Point to existing SDAMPLE YAML file in vSPhere docs for full YAML example.
+
+.Sample VMware vSphere YAML configuration file with failure domain values
+[source,yaml]
+----
+template:
```
[the remark this answers is not in the document]
Comment (rvanderp3) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:19]: I would include a link

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 2 above)
Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: Thanks for the reply. Is the `infrastructure resource specification` the same as the `install-config.yaml` [file](https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installing-vsphere-installer-provisioned-network-customizations.html#installation-installer-provisioned-vsphere-config-yaml_installing-vsphere-installer-provisioned-network-customizations)?

Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: Both the `infrastructure resource specification` and CPMS `providerspec` use name parameters, so the CPMS file is where failure domains gets assigned to a region? Why does the `infrastructure resource specification` spec need to define a name for the failure domain?

File: modules/cpmso-yaml-failure-domain-vsphere.adoc [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// Module included in the following assemblies:
+//
+// * machine_management/cpmso-configuration.adoc
+
+:_mod-docs-content-type: REFERENCE
+[id="cpmso-yaml-failure-domain-vsphere_{context}"]
+= Sample VMware vSphere failure domain configuration
+
+On VMware vSphere infrastructure, the infrastructure resource specification, `infrastructures.config.openshift.io`, defines failure domains for your cluster. The `providerSpec` in the `ControlPlaneMachineSet` custom resource (CR) specifies names for failure domains. A failure domain is an infrastructure resource that comprises a control plane machine set, a vCenter datacenter, vCenter datastore, and a network.
```
Comment (mburke5678) [inline modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: The only place I see `infrastructures.config.openshift.io` with a name, it is called the [cluster-wide infrastructure CRD](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/post-install-vsphere-zones-regions-configuration.html#references-regions-zones-infrastructure-vsphere_post-install-vsphere-zones-regions-configuration). I don't see _infrastructure resource specification_ being used.

File: modules/cpmso-yaml-failure-domain-vsphere.adoc [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+//
+// * machine_management/cpmso-configuration.adoc
+
+:_mod-docs-content-type: REFERENCE
+[id="cpmso-yaml-failure-domain-vsphere_{context}"]
+= Sample VMware vSphere failure domain configuration
+
+On VMware vSphere infrastructure, the infrastructure resource specification, `infrastructures.config.openshift.io`, defines failure domains for your cluster. The `providerSpec` in the `ControlPlaneMachineSet` custom resource (CR) specifies names for failure domains. A failure domain is an infrastructure resource that comprises a control plane machine set, a vCenter datacenter, vCenter datastore, and a network. 
+
+By using a failure domain resource, you can use a control plane machine set to deploy control plane machines on hardware that is separate from the primary VMware vSphere infrastructure. A control plane machine set also balance control plane machines across defined failure domains to provide fault tolerance capabilities to your infrastructure.
```
Comment (mburke5678) [inline modules/cpmso-yaml-failure-domain-vsphere.adoc:11]:
```suggestion
By using a failure domain resource, you can use a control plane machine set to deploy control plane machines on hardware that is separate from the primary VMware vSphere infrastructure. A control plane machine set also balances control plane machines across defined failure domains to provide fault tolerance capabilities to your infrastructure.
```

File: machine_management/control_plane_machine_management/cpmso-getting-started.adoc [hunk 6]
```diff
@@ -63,6 +63,7 @@ The status of the control plane machine set after installation depends on your c
 1. AWS clusters that are upgraded from version 4.11 or earlier require xref:../../machine_management/control_plane_machine_management/cpmso-getting-started.adoc#cpmso-activating_cpmso-getting-started[CR activation].
 2. GCP and Azure clusters that are upgraded from version 4.12 or earlier require xref:../../machine_management/control_plane_machine_management/cpmso-getting-started.adoc#cpmso-activating_cpmso-getting-started[CR activation].
 3. Nutanix and {rh-openstack} clusters that are upgraded from version 4.13 or earlier require xref:../../machine_management/control_plane_machine_management/cpmso-getting-started.adoc#cpmso-activating_cpmso-getting-started[CR activation].
+4. In {product-title} {product-version}, installing a cluster with an active generated CR on VWware vSphere is available as a link:https://access.redhat.com/support/offerings/techpreview[Technology Preview] feature. To enable the feature, set the `featureSet` parameter to `TechPreviewNoUpgrade` in the `install-config.yaml file`.
```
Comment (mburke5678) [inline machine_management/control_plane_machine_management/cpmso-getting-started.adoc:66]: In the preview, the {product-title} here is not rendering.

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 4 above)
Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: I confirmed with the SME that both names are valid,but out of consistency I'll update to `[cluster-wide infrastructure CRD](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/post-install-vsphere-zones-regions-configuration.html#references-regions-zones-infrastructure-vsphere_post-install-vsphere-zones-regions-configuration)` after you finish your review. Thanks!

File: modules/cpmso-yaml-failure-domain-vsphere.adoc [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      - name: <failure_domain_name1>
+      - name: <failure_domain_name2> 
+#...    
+----
+<1> A failure domain defines the vCenter location for {product-title} cluster nodes.
+<2> Defines failure domains by name for the control plane machine set.
+
+[IMPORTANT]
+====
+Each `failureDomains.platform.vsphere.name` field value in the `ControlPlaneMachineSet` CR must match the corresponding value defined in the `failureDomains.name` field of the infrastructure resource specification. Currently, the `vsphere.name` field is the only supported failure domain field that you can specify in the `ControlPlaneMachineSet` CR.
```
Comment (mburke5678) [inline modules/cpmso-yaml-failure-domain-vsphere.adoc:40]: I don't see infrastructure resource specification being used in the docs.

File: machine_management/control_plane_machine_management/cpmso-configuration.adoc [hunk 8]
```diff
@@ -97,6 +97,13 @@ Some sections of the control plane machine set CR are provider-specific. The fol
 //Sample VMware vSphere provider specification
 include::modules/cpmso-yaml-provider-spec-vsphere.adoc[leveloffset=+2]
 
+//Sample VMware vSphere failure domain configuration
+include::modules/cpmso-yaml-failure-domain-vsphere.adoc[leveloffset=+2]
+
+[role="_additional-resources"]
+.Additional resources
+* For an example of an infrastructure resource specification that defines resources for each failure domain, see xref:../../post_installation_configuration/post-install-vsphere-zones-regions-configuration.adoc#specifying-regions-zones-infrastructure-vsphere_post-install-vsphere-zones-regions-configuration[Specifying multiple regions and zones for your cluster on vSphere]
```
Comment (mburke5678) [inline machine_management/control_plane_machine_management/cpmso-configuration.adoc:105]: The only place I see `infrastructures.config.openshift.io` with a name, it is called the [cluster-wide infrastructure CRD](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/post-install-vsphere-zones-regions-configuration.html#references-regions-zones-infrastructure-vsphere_post-install-vsphere-zones-regions-configuration). I don't see _infrastructure resource specification_ being used.

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 4 above)
Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: Only instance exists in the openshift-docs repo as far as grep is concerned:
Screenshot from 2024-01-31 17-27-59

File: modules/cpmso-yaml-failure-domain-vsphere.adoc [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+====
+If you modify the `ProviderSpec` configuration in the `ControlPlaneMachineSet` CR, the control plane machine set updates all control plane machines deployed on the primary infrastructure and each failure domain infrastructure.
+====
+
+:FeatureName: Defining a failure domain for a control plane machine set
+include::snippets/technology-preview.adoc[]
+
+.Example ProviderSpec configuration with specified failure domain names
+[source,yaml]
+----
```
Comment (mburke5678) [inline modules/cpmso-yaml-failure-domain-vsphere.adoc:23]: Per the docs guidelines, code blocks for an [object must include the metadata](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#required-fields): 

```
apiVersion: machine.openshift.io/v1
kind: ControlPlaneMachineSet
metadata:
  name: cluster
  namespace: openshift-machine-api
# ...
```

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 4 above)
Comment (mburke5678) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: Yeah, I see it twice in that module. Where the module is titled _[Parameters for the cluster-wide infrastructure CRD](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/post-install-vsphere-zones-regions-configuration.html#references-regions-zones-infrastructure-vsphere_post-install-vsphere-zones-regions-configuration)_, I thought that might be the winning entry.

Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:9]: :heavy_check_mark:

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 5 above)
Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:11]: :heavy_check_mark:

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 7 above)
Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:40]: :heavy_check_mark:

File: machine_management/control_plane_machine_management/cpmso-configuration.adoc (hunk 8 above)
Comment (dfitzmau) [reply machine_management/control_plane_machine_management/cpmso-configuration.adoc:105]: :heavy_check_mark:

File: modules/cpmso-yaml-failure-domain-vsphere.adoc (hunk 9 above)
Comment (dfitzmau) [reply modules/cpmso-yaml-failure-domain-vsphere.adoc:23]: Good to know. Updated and I'll be mindful in  future. @rvanderp3, does the metadata YAML lines LGTY with regards to values and indentation?

File: machine_management/control_plane_machine_management/cpmso-getting-started.adoc (hunk 6 above)
Comment (dfitzmau) [reply machine_management/control_plane_machine_management/cpmso-getting-started.adoc:66]: The preview links do not update in sync with the new build. Attribute renders OK.

**Comment by bscott-rh** [comment]
Merge review LGTM.

**Pull request merged**